File: robocode-tankroyale-sample-bots/dotnet/TrackFire/src/TrackFire.cs
using System;
using Robocode.TankRoyale.BotApi.Events;

namespace Robocode.TankRoyale.BotApi.Sample.Bots
{
  /// <summary>
  /// TrackFire - a sample bot, original version by Mathew Nelson for Robocode.
  /// Modified by Flemming N. Larsen.
  /// 
  /// Sits still. Tracks and fires at the nearest robot it sees.
  /// </summary>
  public class TrackFire : Bot
  {
    bool isScanning; // flag set when scanning

    // Main method starts our bot
    static void Main(string[] args)
    {
      new TrackFire().Start();
    }

    // Constructor, which loads the bot settings file
    TrackFire() : base(BotInfo.FromJsonFile("trackfire-settings.json")) { }

    // TrackFire's run method
    public override void Run()
    {
      isScanning = false; // reset scanning flag

      // Set colors
      var pink = "#FF69B4";
      SetBodyColor(pink);
      SetTurretColor(pink);
      SetRadarColor(pink);
      SetScanColor(pink);
      SetBulletColor(pink);

      // Loop while running
      while (IsRunning)
      {
        if (isScanning)
          Go(); // skip turn if we a scanning
        else
          TurnGunLeft(10); // Scans automatically as radar is mounted on gun
      }
    }

    // OnScannedRobot: We have a target. Go get it.
    public override void OnScannedBot(ScannedBotEvent e)
    {
      isScanning = true; // we started scanning

      // Calculate direction of the scanned bot and bearing to it for the gun
      double bearingFromGun = GunBearingTo(e.X, e.Y);

      // Turn the gun toward the scanned bot
      TurnGunLeft(bearingFromGun);

      // If it is close enough, fire!
      if (Math.Abs(bearingFromGun) <= 3 && GunHeat == 0)
      {
        Fire(Math.Min(3 - Math.Abs(bearingFromGun), Energy - .1));
      }

      // Generates another scan event if we see a robot.
      // We only need to call this if the gun (and therefore radar)
      // are not turning. Otherwise, scan is called automatically.
      if (bearingFromGun < 5)
        Scan();

      isScanning = false; // we stopped scanning
    }

    // OnWonRound: Do a victory dance!
    public override void OnWonRound(WonRoundEvent e)
    {
      // Victory dance turning right 360 degrees 100 times
      TurnLeft(36_000);
    }
  }
}

File: tank-royale/sample-bots/nim/TrackFire/TrackFire.nim
## TrackFire
##
## A sample bot originally made for Robocode by Mathew Nelson.
## Ported to Nim for Robocode Tank Royale.
##
## Sits still while tracking and firing at the nearest robot it
## detects.

import asyncdispatch
import math
import robocode_tank_royale

type
  TrackFire = ref object of Bot

method run(self: TrackFire) {.async.} =
  ## Called when a new round is started -> initialize and do some movement
  ## Set colors
  var pink = Color.fromRgb(0xFF, 0x69, 0xB4)
  await self.setBodyColor(pink)
  await self.setTurretColor(pink)
  await self.setRadarColor(pink)
  await self.setScanColor(pink)
  await self.setBulletColor(pink)

  ## Loop while running
  while self.isRunning():
    await self.turnGunRight(10.0) ## Scans automatically as radar is mounted on gun

method onScannedBot(self: TrackFire, event: ScannedBotEvent) {.async.} =
  ## We scanned another bot -> we have a target, so go get it
  ## Calculate direction of the scanned bot and bearing to it for the gun
  var bearingFromGun = self.gunBearingTo(event.x, event.y)

  ## Turn the gun toward the scanned bot
  await self.turnGunLeft(bearingFromGun)

  ## If it is close enough, fire!
  if abs(bearingFromGun) <= 3.0 and self.getGunHeat() == 0.0:
    await self.fire(min(3.0 - abs(bearingFromGun), self.getEnergy() - 0.1))

  ## Rescan immediately to keep tracking the target bot
  await self.rescan()

method onWonRound(self: TrackFire, event: WonRoundEvent) {.async.} =
  ## We won the round -> do a victory dance!
  ## Victory dance turning right 360 degrees 100 times
  await self.turnRight(36000.0)

when isMainModule:
  let bot = TrackFire()
  waitFor bot.start()